ring.go

```diff
@@ -213,32 +215,54 @@ type ringShards struct {
 
 	mu       sync.RWMutex
 	hash     ConsistentHash
-	shards   map[string]*ringShard // read only
-	list     []*ringShard          // read only
+	shards   map[string]*ringShard // read only, updated by SetAddrs
+	list     []*ringShard          // read only, updated by SetAddrs
 	numShard int
 	closed   bool
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
-	list := make([]*ringShard, 0, len(shards))
+	c := &ringShards{
+		opt: opt,
+	}
+	c.SetAddrs(opt.Addrs)
 
-	for name, addr := range opt.Addrs {
-		shard := newRingShard(opt, name, addr)
-		shards[name] = shard
+	return c
+}
 
-		list = append(list, shard)
+func (c *ringShards) SetAddrs(addrs map[string]string) {
```

Comment:
I think we may need to protect it from concurrent calls

Reply:
Do you mean concurrent calls to SetAddrs?

Reply:
added godoc